File: chapechape_partner/lib/core/services/api/audit_service.dart
import 'package:dio/dio.dart';
import 'package:flutter/foundation.dart';
import '../../../core/config/app_config_manager.dart';

/// Service pour gérer l'audit et la sécurité
class AuditService {
  final Dio _dio;

  AuditService(this._dio);

  /// Obtenir l'historique de sécurité d'un utilisateur
  Future<SecurityHistoryResult> getSecurityHistory({int limit = 50}) async {
    try {
      final response = await _dio.get(
        '/audit/security-history',
        queryParameters: {'limit': limit},
      );

      if (response.statusCode == 200 && response.data['success'] == true) {
        final List<dynamic> historyData = response.data['data'] ?? [];
        final history = historyData.map((item) => SecurityActivity.fromJson(item)).toList();

        return SecurityHistoryResult(
          success: true,
          history: history,
        );
      } else {
        return SecurityHistoryResult(
          success: false,
          message: response.data['message'] ?? 'Erreur lors de la récupération de l\'historique',
        );
      }
    } catch (e) {
      debugPrint('❌ Erreur getSecurityHistory: $e');
      return SecurityHistoryResult(
        success: false,
        message: 'Erreur de connexion: $e',
      );
    }
  }

  /// Obtenir les statistiques de sécurité
  Future<SecurityStatsResult> getSecurityStats({int days = 30}) async {
    try {
      final response = await _dio.get(
        '/audit/security-stats',
        queryParameters: {'days': days},
      );

      if (response.statusCode == 200 && response.data['success'] == true) {
        final statsData = response.data['data'];
        final stats = SecurityStats.fromJson(statsData);

        return SecurityStatsResult(
          success: true,
          stats: stats,
        );
      } else {
        return SecurityStatsResult(
          success: false,
          message: response.data['message'] ?? 'Erreur lors de la récupération des statistiques',
        );
      }
    } catch (e) {
      debugPrint('❌ Erreur getSecurityStats: $e');
      return SecurityStatsResult(
        success: false,
        message: 'Erreur de connexion: $e',
      );
    }
  }

  /// Obtenir le journal d'activité complet
  Future<ActivityLogResult> getActivityLog({
    int page = 1,
    int limit = 20,
    String? module,
    String? action,
    String? severity,
  }) async {
    try {
      final queryParams = <String, dynamic>{
        'page': page,
        'limit': limit,
      };

      if (module != null) queryParams['module'] = module;
      if (action != null) queryParams['action'] = action;
      if (severity != null) queryParams['severity'] = severity;

      final response = await _dio.get(
        '/audit/activity-log',
        queryParameters: queryParams,
      );

      if (response.statusCode == 200 && response.data['success'] == true) {
        final data = response.data['data'];
        final List<dynamic> activitiesData = data['activities'] ?? [];
        final activities = activitiesData.map((item) => SecurityActivity.fromJson(item)).toList();
        
        final paginationData = data['pagination'];
        final pagination = PaginationInfo.fromJson(paginationData);

        return ActivityLogResult(
          success: true,
          activities: activities,
          pagination: pagination,
        );
      } else {
        return ActivityLogResult(
          success: false,
          message: response.data['message'] ?? 'Erreur lors de la récupération du journal',
        );
      }
    } catch (e) {
      debugPrint('❌ Erreur getActivityLog: $e');
      return ActivityLogResult(
        success: false,
        message: 'Erreur de connexion: $e',
      );
    }
  }
}

/// Modèle pour une activité de sécurité
class SecurityActivity {
  final String id;
  final String action;
  final String module;
  final String description;
  final String ipAddress;
  final String? userAgent;
  final LocationInfo? location;
  final DeviceInfo? device;
  final String status;
  final String severity;
  final int riskScore;
  final bool isSuspicious;
  final DateTime createdAt;
  final Map<String, dynamic> metadata;

  SecurityActivity({
    required this.id,
    required this.action,
    required this.module,
    required this.description,
    required this.ipAddress,
    this.userAgent,
    this.location,
    this.device,
    required this.status,
    required this.severity,
    required this.riskScore,
    required this.isSuspicious,
    required this.createdAt,
    required this.metadata,
  });

  factory SecurityActivity.fromJson(Map<String, dynamic> json) {
    return SecurityActivity(
      id: json['_id'] ?? '',
      action: json['action'] ?? '',
      module: json['module'] ?? '',
      description: json['description'] ?? '',
      ipAddress: json['ipAddress'] ?? '',
      userAgent: json['userAgent'],
      location: json['location'] != null ? LocationInfo.fromJson(json['location']) : null,
      device: json['device'] != null ? DeviceInfo.fromJson(json['device']) : null,
      status: json['status'] ?? '',
      severity: json['severity'] ?? '',
      riskScore: json['riskScore'] ?? 0,
      isSuspicious: json['isSuspicious'] ?? false,
      createdAt: DateTime.parse(json['createdAt']),
      metadata: Map<String, dynamic>.from(json['metadata'] ?? {}),
    );
  }

  /// Obtenir l'icône selon l'action
  String get actionIcon {
    switch (action) {
      case 'login':
        return '🔐';
      case 'logout':
        return '🚪';
      case 'login_failed':
        return '❌';
      case 'password_change':
        return '🔑';
      case 'email_change':
        return '📧';
      case 'phone_change':
        return '📱';
      case 'bank_account_change':
        return '🏦';
      case 'payout_initiated':
        return '💸';
      case 'security_alert':
        return '⚠️';
      default:
        return '📝';
    }
  }

  /// Obtenir la couleur selon la gravité
  String get severityColor {
    switch (severity) {
      case 'low':
        return 'green';
      case 'medium':
        return 'orange';
      case 'high':
        return 'red';
      case 'critical':
        return 'purple';
      default:
        return 'grey';
    }
  }
}

/// Informations de localisation
class LocationInfo {
  final String country;
  final String region;
  final String city;

  LocationInfo({
    required this.country,
    required this.region,
    required this.city,
  });

  factory LocationInfo.fromJson(Map<String, dynamic> json) {
    return LocationInfo(
      country: json['country'] ?? 'Unknown',
      region: json['region'] ?? 'Unknown',
      city: json['city'] ?? 'Unknown',
    );
  }
}

/// Informations de l'appareil
class DeviceInfo {
  final String type;
  final String os;
  final String browser;

  DeviceInfo({
    required this.type,
    required this.os,
    required this.browser,
  });

  factory DeviceInfo.fromJson(Map<String, dynamic> json) {
    return DeviceInfo(
      type: json['type'] ?? 'unknown',
      os: json['os'] ?? 'unknown',
      browser: json['browser'] ?? 'unknown',
    );
  }
}

/// Statistiques de sécurité
class SecurityStats {
  final int totalActivities;
  final int suspiciousActivities;
  final int failedLogins;
  final int highRiskActivities;
  final double averageRiskScore;

  SecurityStats({
    required this.totalActivities,
    required this.suspiciousActivities,
    required this.failedLogins,
    required this.highRiskActivities,
    required this.averageRiskScore,
  });

  factory SecurityStats.fromJson(Map<String, dynamic> json) {
    return SecurityStats(
      totalActivities: json['totalActivities'] ?? 0,
      suspiciousActivities: json['suspiciousActivities'] ?? 0,
      failedLogins: json['failedLogins'] ?? 0,
      highRiskActivities: json['highRiskActivities'] ?? 0,
      averageRiskScore: (json['averageRiskScore'] ?? 0).toDouble(),
    );
  }
}

/// Informations de pagination
class PaginationInfo {
  final int page;
  final int limit;
  final int total;
  final int pages;

  PaginationInfo({
    required this.page,
    required this.limit,
    required this.total,
    required this.pages,
  });

  factory PaginationInfo.fromJson(Map<String, dynamic> json) {
    return PaginationInfo(
      page: json['page'] ?? 1,
      limit: json['limit'] ?? 20,
      total: json['total'] ?? 0,
      pages: json['pages'] ?? 1,
    );
  }
}

/// Résultat de l'historique de sécurité
class SecurityHistoryResult {
  final bool success;
  final String? message;
  final List<SecurityActivity> history;

  SecurityHistoryResult({
    required this.success,
    this.message,
    this.history = const [],
  });
}

/// Résultat des statistiques de sécurité
class SecurityStatsResult {
  final bool success;
  final String? message;
  final SecurityStats? stats;

  SecurityStatsResult({
    required this.success,
    this.message,
    this.stats,
  });
}

/// Résultat du journal d'activité
class ActivityLogResult {
  final bool success;
  final String? message;
  final List<SecurityActivity> activities;
  final PaginationInfo? pagination;

  ActivityLogResult({
    required this.success,
    this.message,
    this.activities = const [],
    this.pagination,
  });
}

File: backend/src/services/audit.service.js
const ActivityLog = require('../models/activityLog.model');
const LoginAttempt = require('../models/loginAttempt.model');
const BlockedIP = require('../models/blockedIP.model');
const User = require('../models/user.model');
const logger = require('../utils/logger');
const axios = require('axios');

/**
 * Service d'audit et de sécurité avancé
 * Gère la journalisation des activités sensibles et la détection d'activités suspectes
 */
class AuditService {
    constructor() {
        this.suspiciousPatterns = {
            multipleFailedLogins: 5, // Nombre de tentatives échouées avant alerte
            unusualLocation: true,   // Détecter les connexions depuis des lieux inhabituels
            rapidActions: 10,        // Nombre d'actions rapides avant alerte
            highRiskActions: ['password_change', 'bank_account_change', 'payout_initiated']
        };
    }

    /**
     * Enregistrer une activité dans l'audit trail
     * @param {Object} params - Paramètres de l'activité
     */
    async logActivity({
        userId,
        action,
        module,
        description,
        ipAddress,
        userAgent,
        metadata = {},
        status = 'success',
        severity = 'low'
    }) {
        try {
            // Calculer le score de risque
            const riskScore = await this.calculateRiskScore({
                userId,
                action,
                ipAddress,
                userAgent,
                metadata
            });

            // Détecter les activités suspectes
            const isSuspicious = await this.detectSuspiciousActivity({
                userId,
                action,
                ipAddress,
                riskScore
            });

            // Obtenir les informations de localisation
            const location = await this.getLocationInfo(ipAddress);

            // Analyser l'appareil
            const device = this.analyzeDevice(userAgent);

            const activityLog = new ActivityLog({
                user: userId,
                action,
                module,
                description,
                ipAddress,
                userAgent,
                location,
                device,
                metadata,
                status,
                severity,
                riskScore,
                isSuspicious
            });

            await activityLog.save();

            // Si activité suspecte, déclencher une alerte
            if (isSuspicious) {
                await this.triggerSecurityAlert(userId, action, ipAddress, riskScore);
            }

            logger.info(`Activity logged: ${action} for user ${userId}`, {
                action,
                userId,
                riskScore,
                isSuspicious
            });

            return activityLog;
        } catch (error) {
            logger.error('Error logging activity:', error);
            throw error;
        }
    }

    /**
     * Calculer le score de risque d'une activité
     */
    async calculateRiskScore({ userId, action, ipAddress, userAgent, metadata }) {
        let riskScore = 0;

        // Actions à haut risque
        if (this.suspiciousPatterns.highRiskActions.includes(action)) {
            riskScore += 30;
        }

        // Vérifier les tentatives de connexion échouées récentes
        const recentFailedLogins = await LoginAttempt.countDocuments({
            ip: ipAddress,
            success: false,
            lastAttempt: { $gte: new Date(Date.now() - 24 * 60 * 60 * 1000) }
        });

        if (recentFailedLogins >= 3) {
            riskScore += 25;
        }

        // Vérifier si l'IP est bloquée
        const isBlocked = await BlockedIP.findOne({ ip: ipAddress });
        if (isBlocked) {
            riskScore += 50;
        }

        // Vérifier les activités récentes de l'utilisateur
        const recentActivities = await ActivityLog.countDocuments({
            user: userId,
            createdAt: { $gte: new Date(Date.now() - 60 * 60 * 1000) } // Dernière heure
        });

        if (recentActivities > 20) {
            riskScore += 20;
        }

        // Vérifier les changements de localisation
        const lastActivity = await ActivityLog.findOne({
            user: userId,
            ipAddress: { $ne: ipAddress }
        }).sort({ createdAt: -1 });

        if (lastActivity) {
            const timeDiff = Date.now() - lastActivity.createdAt.getTime();
            if (timeDiff < 60 * 60 * 1000) { // Moins d'une heure
                riskScore += 15;
            }
        }

        return Math.min(riskScore, 100);
    }

    /**
     * Détecter les activités suspectes
     */
    async detectSuspiciousActivity({ userId, action, ipAddress, riskScore }) {
        // Score de risque élevé
        if (riskScore >= 70) {
            return true;
        }

        // Actions critiques
        if (['password_change', 'bank_account_change', 'payout_initiated'].includes(action)) {
            return riskScore >= 40;
        }

        // Connexions depuis des IP inhabituelles
        const userActivities = await ActivityLog.find({
            user: userId,
            action: 'login',
            ipAddress: { $ne: ipAddress }
        }).sort({ createdAt: -1 }).limit(5);

        if (userActivities.length >= 3) {
            return true;
        }

        return false;
    }

    /**
     * Obtenir les informations de localisation depuis l'IP
     */
    async getLocationInfo(ipAddress) {
        try {
            const response = await axios.get(
                `http://ip-api.com/json/${ipAddress}?fields=country,regionName,city`,
                { timeout: 5000 }
            );
            
            if (response.data && response.data.country) {
                return {
                    country: response.data.country,
                    region: response.data.regionName,
                    city: response.data.city
                };
            }
        } catch (error) {
            logger.warn('Failed to get location info:', error.message);
        }

        return {
            country: 'Unknown',
            region: 'Unknown',
            city: 'Unknown'
        };
    }

    /**
     * Analyser les informations de l'appareil
     */
    analyzeDevice(userAgent) {
        if (!userAgent) {
            return { type: 'unknown', os: 'unknown', browser: 'unknown' };
        }

        const ua = userAgent.toLowerCase();
        
        // Type d'appareil
        let type = 'desktop';
        if (ua.includes('mobile') || ua.includes('android') || ua.includes('iphone')) {
            type = 'mobile';
        } else if (ua.includes('tablet') || ua.includes('ipad')) {
            type = 'tablet';
        }

        // OS
        let os = 'unknown';
        if (ua.includes('windows')) os = 'Windows';
        else if (ua.includes('mac')) os = 'macOS';
        else if (ua.includes('linux')) os = 'Linux';
        else if (ua.includes('android')) os = 'Android';
        else if (ua.includes('ios') || ua.includes('iphone')) os = 'iOS';

        // Navigateur
        let browser = 'unknown';
        if (ua.includes('chrome')) browser = 'Chrome';
        else if (ua.includes('firefox')) browser = 'Firefox';
        else if (ua.includes('safari')) browser = 'Safari';
        else if (ua.includes('edge')) browser = 'Edge';

        return { type, os, browser };
    }

    /**
     * Déclencher une alerte de sécurité
     */
    async triggerSecurityAlert(userId, action, ipAddress, riskScore) {
        try {
            const user = await User.findById(userId);
            if (!user) return;

            // Enregistrer l'alerte
            await this.logActivity({
                userId,
                action: 'security_alert',
                module: 'security',
                description: `Alerte de sécurité: ${action} depuis ${ipAddress}`,
                ipAddress,
                userAgent: 'System',
                metadata: { riskScore, originalAction: action },
                status: 'warning',
                severity: riskScore >= 80 ? 'critical' : 'high'
            });

            // Envoyer notification au partenaire
            const notificationService = require('./notification.service');
            await notificationService.notifySecurityAlert(
                userId,
                `Alerte de sécurité: Activité suspecte détectée`,
                {
                    action,
                    ipAddress,
                    riskScore,
                    timestamp: new Date()
                }
            );

            logger.warn(`Security alert triggered for user ${userId}: ${action}`, {
                userId,
                action,
                ipAddress,
                riskScore
            });
        } catch (error) {
            logger.error('Error triggering security alert:', error);
        }
    }

    /**
     * Obtenir l'historique de sécurité d'un utilisateur
     */
    async getSecurityHistory(userId, limit = 50) {
        try {
            const activities = await ActivityLog.find({
                user: userId,
                $or: [
                    { action: { $in: ['login', 'logout', 'login_failed', 'password_change', 'security_alert'] } },
                    { isSuspicious: true },
                    { severity: { $in: ['high', 'critical'] } }
                ]
            })
            .sort({ createdAt: -1 })
            .limit(limit)
            .populate('user', 'firstName lastName email');

            return activities;
        } catch (error) {
            logger.error('Error getting security history:', error);
            throw error;
        }
    }

    /**
     * Obtenir les statistiques de sécurité
     */
    async getSecurityStats(userId, days = 30) {
        try {
            const startDate = new Date(Date.now() - days * 24 * 60 * 60 * 1000);

            const stats = await ActivityLog.aggregate([
                {
                    $match: {
                        user: userId,
                        createdAt: { $gte: startDate }
                    }
                },
                {
                    $group: {
                        _id: null,
                        totalActivities: { $sum: 1 },
                        suspiciousActivities: {
                            $sum: { $cond: [{ $eq: ['$isSuspicious', true] }, 1, 0] }
                        },
                        failedLogins: {
                            $sum: { $cond: [{ $eq: ['$action', 'login_failed'] }, 1, 0] }
                        },
                        highRiskActivities: {
                            $sum: { $cond: [{ $in: ['$severity', ['high', 'critical']] }, 1, 0] }
                        },
                        averageRiskScore: { $avg: '$riskScore' }
                    }
                }
            ]);

            return stats[0] || {
                totalActivities: 0,
                suspiciousActivities: 0,
                failedLogins: 0,
                highRiskActivities: 0,
                averageRiskScore: 0
            };
        } catch (error) {
            logger.error('Error getting security stats:', error);
            throw error;
        }
    }
}

module.exports = new AuditService();


